Forwarded: yes
Author: Dmitry Smirnov <onlyjob@member.fsf.org>
Description: rbdmap: lazy umount on shutdown and reboot (runlevel 0 and 6)

Do lazy unmount in runlevel 0 and 6. This is necessary to avoid unclean
shutdown due to "umount" hangs on reboot.

"rbdmap" fail to umount and unmap RBD device when some applications are
still using file system on RBD. If RBD device not released on
shutdown/reboot system deconfigures network interfaces, terminates
remaining processes then tries to umount remaining file systems where it
hangs forever in endless libceph attempt to reach MONs. This scenario was
observed when /home is located on RBD device (users start processes in
screen/tmux etc.). Even worse, `umount` stuck on RBD mount point so
remaining local file systems are never un-mounted making unclean shutdown
almost inevitable.

Lazy umount is effective because it allows to release RBD device and
unmount file system when applications are terminated.

Also before umount (if `fuser` utility is available) TERM signal will be
sent to all processes still using mount point so applications that haven't
been asked to terminate yet will have some time to finish before final cut.

Signed-off-by: Dmitry Smirnov <onlyjob@member.fsf.org>
---
 src/init-rbdmap | 13 ++++++++++++-
 1 file changed, 12 insertions(+), 1 deletion(-)

--- a/src/init-rbdmap
+++ b/src/init-rbdmap
@@ -105,7 +105,18 @@ do_unmap() {
 			MNT=$(findmnt --mtab --source ${DEV} --noheadings | awk '{print $1'})
 			if [ -n "${MNT}" ]; then
 			    logger -p "daemon.debug" -t init-rbdmap "Unmounting '${MNT}'"
-			    umount "${MNT}" >>/dev/null 2>&1
+			    local ULAZY=""
+			    local RUNLEVEL=$(runlevel | awk '{print $2}')
+			    if [ $? -eq 0 ]; then
+			        if [ "${RUNLEVEL}" -eq 0 ] \
+			        || [ "${RUNLEVEL}" -eq 6 ] ; then
+			            ULAZY="-l"
+			            if [ -x "$(which fuser)" ]; then
+			                fuser -M -m "${MNT}" --kill -TERM
+			            fi
+			        fi
+			    fi
+			    umount ${ULAZY} "${MNT}" >>/dev/null 2>&1
 			fi
 			if mountpoint -q "${MNT}"; then
 			    ## Un-mounting failed.
